src/font.rs

```diff
@@ -114,6 +114,273 @@ impl<'a> Loca<'a> {
+            result.push(get_i16(data, vec_position as usize).unwrap());
+            vec_position += 2;
+        }
+        result
+    }
+
+    fn get_end_counts_position() -> u16 {
+        14
+    }
+
+    fn get_end_counts(&'a self) -> Vec<u16> {
+        let seg_count = self.get_seg_count();
+        self.get_u16_vec(Self::get_end_counts_position(), seg_count)
+    }
+
+    fn get_start_counts_position(seg_count: u16) -> u16 {
+        Self::get_end_counts_position() + 2 + 2 * seg_count
+    }
+
+    fn get_start_counts(&'a self) -> Vec<u16> {
+        let seg_count = self.get_seg_count();
+        self.get_u16_vec(Self::get_start_counts_position(seg_count), seg_count)
+    }
+
+    fn get_id_deltas_position(seg_count: u16) -> u16 {
+        Self::get_start_counts_position(seg_count) + 2 * seg_count
+    }
+
+    fn get_id_deltas(&'a self) -> Vec<i16> {
+        let seg_count = self.get_seg_count();
+        self.get_i16_vec(Self::get_id_deltas_position(seg_count), seg_count)
+    }
+
+    fn get_id_range_offset_position(seg_count: u16) -> u16 {
+        Self::get_id_deltas_position(seg_count) + 2 * seg_count
+    }
+
+    fn get_id_range_offsets(&'a self) -> Vec<u16> {
+        let seg_count = self.get_seg_count();
+        self.get_u16_vec(Self::get_id_range_offset_position(seg_count), seg_count)
+    }
+
+    pub fn lookup_glyph_id(&'a self, code_point: u16) -> Option<u16> {
+        if self.get_format() != 4 {
+            return None;
+        }
+        let &Encoding(data) = self;
+        let end_counts_position = Self::get_end_counts_position();
+        let seg_count = self.get_seg_count();
+        let mut size = seg_count - 1;
+        let mut index = size / 2;
+        while size > 0 {
+            let search = end_counts_position + index * 2;
+            let end_value = get_u16(data, search as usize).unwrap();
+            if end_value >= code_point {
+                let start_pos = Self::get_start_counts_position(seg_count) + 2 * index;
+                let start_value = get_u16(data, start_pos as usize).unwrap();
+                if start_value > code_point {
+                    size /= 2;
+                    index -= size;
+                } else {
+                    let id_range_offset_pos =
+                        Self::get_id_range_offset_position(seg_count) + 2 * index;
+                    let id_range_offset_value =
+                        get_u16(data, id_range_offset_pos as usize).unwrap();
+                    let id_delta_pos = Self::get_id_deltas_position(seg_count) + 2 * index;
+                    let id_delta = get_i16(data, id_delta_pos as usize).unwrap();
+                    return if id_range_offset_value == 0 {
+                        Some((code_point as i16 + id_delta) as u16)
+                    } else {
+                        let delta = (code_point - start_value) * 2;
+                        let pos = id_range_offset_pos + delta + id_range_offset_value;
```

Comment:
I haven't researched this carefully, but I suspect you need wrapping_add for this, I believe it can wrap in some cases.

Reply:
I think you are right in both cases.